hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyRatisContainerReplicaCount.java

```diff
@@ -59,4 +65,34 @@ protected int healthyReplicaCountAdapter() {
   public int getUnhealthyReplicaCountAdapter() {
     return getMisMatchedReplicaCount();
   }
+
+  @Override
+  public boolean isHealthyEnoughForOffline() {
+    LifeCycleState containerState = getContainer().getState();
+    return (containerState == LifeCycleState.CLOSED
+        || containerState == LifeCycleState.QUASI_CLOSED)
+        && getReplicas().stream()
+        .filter(r -> r.getDatanodeDetails().getPersistedOpState() == IN_SERVICE)
+        .filter(r -> r.getState() !=
+            ContainerReplicaProto.State.UNHEALTHY)
+        .allMatch(r -> ReplicationManager.compareState(
```

Comment:
allMatch:

> If the stream is empty then true is returned and the predicate is not evaluated.

What is the intention if there are only out of service nodes or all are unhealthy? Do we expect this to return true or false?

Reply:
When all are UNHEALTHY, then I think it's fine to return true, because RM will still try to maintain replication factor replicas of that container and we don't want all UNHEALTHY blocking decommissioning. 

Not so sure about the case when all are out of service. Such a case is unexpected because if RM is working properly, it shouldn't allow this to happen in the first place. I think the user of this class should use the other API, `isSufficientlyReplicatedForOffline`, to guard against this case. It seems to be out of scope for `isHealthyEnoughForOffline`, since this method is concerned with the health state of the replicas, while the other one checks for number of replicas. We can still return false from this method just to be defensive, though.